### PR TITLE
feat(cmd/influx): influx cli can write CSV files

### LINKED ISSUE #17003: Add the ability to import an annotated Flux CSV response via the influx cli
This should be the signature of the command:
```
#Import annotated Flux CSV response
influx write --bucket fly --org ACME --format csv --file /path/to/flights.csv
```

```
➜  influxdb git:(master) ✗ ./bin/darwin/influx write -h
Write data to InfluxDB via stdin, or add an entire file specified with the -f flag

Usage:
  influx write [flags]

Flags:
  -b, --bucket string      The name of destination bucket; Maps to env var $INFLUX_BUCKET_NAME
      --bucket-id string   The ID of destination bucket; Maps to env var $INFLUX_BUCKET_ID
  -h, --help               Help for the write command
  -o, --org string         The name of the organization that owns the bucket; Maps to env var $INFLUX_ORG
      --org-id string      The ID of the organization that owns the bucket; Maps to env var $INFLUX_ORG_ID
  -p, --precision string   Precision of the timestamps of the lines; Maps to env var $INFLUX_PRECISION (default "ns")
      --format string      The format of the file to load. Supports lp (Line Protocol) or csv (Comma Separated Values). Defaults based on the file extension, or lp if it cannot be determined.
  -f, --file string        The path to the file to import

Global Flags:
      --host string    HTTP address of Influx; Maps to env var $INFLUX_HOST (default "http://localhost:9999")
      --local          Run commands locally against the filesystem
      --skip-verify    SkipVerify controls whether a client verifies the server's certificate chain and host name.
  -t, --token string   API token to be used throughout client calls; Maps to env var $INFLUX_TOKEN
```

It should be able to load the following csv: 
[2020-02-25-14-22_chronograf_data.csv.zip](https://github.com/influxdata/influxdb/files/4252466/2020-02-25-14-22_chronograf_data.csv.zip)

### PR DESCRIPTION
**Pull request by sranka**
Closes #17003 #17356

This PR 
- supports CSV files in 'influx write'
   - transformation of CSV data to line protocol is driven by CSV annotations
   - annotations described https://v2.docs.influxdata.com/v2.0/reference/syntax/annotated-csv/#annotations are all supported, so that flux query result can be written "as-is"
   - *#datatype* annotation is enhanced:
      * so you can mark non-field data with: 
          * _measurement_, _tag_, _time_
          * _ignore_ value in a column simply ignores the column
          * _dateTime_ is either as a long (int64) number or RFC3339 format, this type is always used to represent time of measurement; or you can specify
             * _dateTime:RFC3339_ for RFC3339 time
             * _dateTime:number_ to expect a long number
          * _time_ value is also supported as an alias for the existing _dateTime_
      * a _field_ data type can be used to let detect field's data type
      * default datatype for a column is =field= unless _field column is present (ignored then)
      * there can be at most 1 _dateTime_ column
   - https://github.com/bonitoo-io/influxdb-csv-import/blob/master/README.md contains demonstrative CSV examples
- adds new options to 'influx write":
```
  -f, --file string        The path to the file to import
      --format string      Input format, either lp (Line Protocol) or csv (Comma Separated Values). Defaults to lp unless '.csv' extension
```
- a new "write dryrun" (i.e. 'write' command with 'dryrun' sub-command) can be used to validate and tune annotations in CSV files; parameters are the as in "write", but line protocol is written to stdout
- keeps the existing functionality (writing line protocol data) without changes
  - the default input format is lp (line protocol) unless a file with ".csv" extension is supplied

- [x] [CHANGELOG.md](https://github.com/influxdata/influxdb/blob/master/CHANGELOG.md) updated with a link to the PR (not the Issue)
- [x] [Well-formatted commit messages](https://www.conventionalcommits.org/en/v1.0.0-beta.3/)
- [x] Rebased/mergeable
- [x] Tests pass
- [x] Documentation updated or issue created (provide link to issue/pr)
- [x] Signed [CLA](https://influxdata.com/community/cla/) (if not already signed)

### CODE REVIEW

File: cmd/influx/main.go [hunk 1]
```diff
@@ -176,12 +176,12 @@ func (b *cmdInfluxBuilder) cmd(childCmdFns ...func(f *globalFlags, opt genericCL
...
-
-		// this is after the flagOpts register b/c we don't want to show the default value
-		// in the usage display. This will add it as the config, then if a token flag
-		// is provided too, the flag will take precedence.
-		flags.Config = getConfigFromDefaultPath()
+		if migrateOldCredential() {
+			// this is after the flagOpts register b/c we don't want to show the default value
+			// in the usage display. This will add it as the config, then if a token flag
+			// is provided too, the flag will take precedence.
+			flags.Config = getConfigFromDefaultPath()
```
Comment (kelwang) [inline cmd/influx/main.go]: This change will break the multiple configs feature. Usage display override is related to the cobra persistent flags, has nothing to do with migration. https://github.com/influxdata/influxdb/issues/17259

Comment (sranka) [reply cmd/influx/main.go]: Understood, I am going to revert this part to a state "before this PR".

File: cmd/influx/main.go [hunk 2]
```diff
@@ -301,6 +301,10 @@ func writeConfigToPath(tok, org, path, dir string) error {
 }
 
 func checkSetup(host string, skipVerify bool) error {
+	if host == "-" {
```
Comment (jsteenb2) [inline cmd/influx/main.go]: Why is this necessary here?

File: cmd/influx/write.go [hunk 3]
```diff
@@ -68,94 +75,128 @@ or add an entire file specified with an @ prefix.`
...
+	if writeFlags.DryRun {
+		flags.Host = "-"
 	}
 
-	if writeFlags.Bucket != "" && writeFlags.BucketID != "" {
-		return fmt.Errorf("please specify one of bucket or bucket-id")
-	}
+	var bucketID, orgID platform.ID
+	// validate flags unless writing to stdout
+	if flags.Host != "-" {
```
Comment (jsteenb2) [inline cmd/influx/write.go]: I don't understand this, why is this here? Maybe a comment added here at the very least

File: cmd/influx/write.go [hunk 4]
```diff
@@ -68,94 +75,128 @@ or add an entire file specified with an @ prefix.`
...
+	}
 
-	s := write.Batcher{
-		Service: &http.WriteService{
-			Addr:               flags.Host,
-			Token:              flags.Token,
-			Precision:          writeFlags.Precision,
-			InsecureSkipVerify: flags.skipVerify,
-		},
+	if writeFlags.Format == inputFormatCsv {
```
Comment (jsteenb2) [inline cmd/influx/write.go]: can we inspect the filename to determine if it is `csv` vs `lp`? if so, we can make this a bit simpler for hte user by not requiring the format, and inspecting the file instead.

File: cmd/influx/write.go [hunk 5]
```diff
@@ -68,94 +75,128 @@ or add an entire file specified with an @ prefix.`
...
-		return err
-	}
+		if !models.ValidPrecision(writeFlags.Precision) {
+			return fmt.Errorf("invalid precision")
+		}
 
-	var filter platform.BucketFilter
-	if writeFlags.BucketID != "" {
-		filter.ID, err = platform.IDFromString(writeFlags.BucketID)
+		bs, err := newBucketService()
```
Comment (kelwang) [inline cmd/influx/write.go]: If you flag dry host to be '-', you could get an error here, since bucket service is a tcp request to the host.

File: cmd/influx/main.go (hunk 2 above)
Comment (sranka) [reply cmd/influx/main.go]: There is a new *--dry-run* flag in "influx write" command that helps to tune CSV to line protocol conversion, it prints lines to stdout. This dry-run does not require any check of influx server setup.

Comment (jsteenb2) [reply cmd/influx/main.go]: have you considered making dry run a new command instead?

```sh
influx write dryrun
```

**Comment by kelwang** [comment]
with this pr, we can also close #17356

File: cmd/influx/write.go [hunk 6]
```diff
@@ -15,20 +15,27 @@ import (
 	"github.com/spf13/cobra"
 )
 
+const (
+	inputFormatCsv          = "csv"
+	inputFormatLineProtocol = "lp"
+)
+
 var writeFlags struct {
 	OrgID     string
```
Comment (jsteenb2) [inline cmd/influx/write.go]: can you replace OrgID and Org with the org type that does this:

https://github.com/influxdata/influxdb/blob/master/cmd/influx/main.go#L360

you can see it being used here:

setting the flags: https://github.com/influxdata/influxdb/blob/master/cmd/influx/pkg.go#L100
validating org id: https://github.com/influxdata/influxdb/blob/master/cmd/influx/pkg.go#L108
getting org id from type: https://github.com/influxdata/influxdb/blob/master/cmd/influx/pkg.go#L122

side note: just found redundant code I wrote in that cmd 😬

File: cmd/influx/write.go (hunk 4 above)
Comment (sranka) [reply cmd/influx/write.go]: The ".csv" file extension now determines the CSV format. We might do better by inspecting the first row of the data that comes in, but such a detection might be error-prone (I don't know yet). I would prefer to do this in a separate issue after this PR is merged.

File: cmd/influx/main.go (hunk 2 above)
Comment (sranka) [reply cmd/influx/main.go]: Thank you, I only wanted to bypass the 'setup' middleware for dry-run, and this solution does not appear to me on the first sight. Will do.

File: cmd/influx/write.go (hunk 3 above)
Comment (sranka) [reply cmd/influx/write.go]: will remove this, and introduce a dry run sub-command

File: cmd/influx/write.go (hunk 6 above)
Comment (sranka) [reply cmd/influx/write.go]: :) ok

File: cmd/influx/write.go (hunk 4 above)
Comment (sranka) [reply cmd/influx/write.go]: Is there any reliable way to detect line protocol on input in the existing code base? Or should we simply assume CSV data if input starts with an annotation comment (no real measurements probably start with #)

If the data input is supplied via stdin, the user should better let us know what data type (lp or csv) comes in ... the user knows, but the tool guesses. I could also imagine users supplying wrong data, from which we will then provide false advice about parsing errors.

@jsteenb2 Can you please comment on this?

Comment (jsteenb2) [reply cmd/influx/write.go]: totally fine breakign this out into separate issue. Its a quality of life feature. I'm not advocating that we remove the format. I was thinking it coudl be optional if we have a file as an input.  I dont' think we have a reliable way to inspect a buffer and pull out if it is LP or csv 🤔. I could be wrong on that. Maybe our storage reviewer (@sebito91) would know.

Comment (jsteenb2) [reply cmd/influx/write.go]: we do something similar to this in the pkg command. We just bet on the file extension being accurate and make decisions off of that if the `encoding` flag isn't set 👍 

https://github.com/influxdata/influxdb/blob/master/cmd/influx/pkg.go#L490

Comment (sranka) [reply cmd/influx/write.go]: Yes, the format flag is already optional herein, defaults to lp (backward compatibility) unless a file with .csv extension supplied.

https://github.com/influxdata/influxdb/blob/17003/writeFromCsv/cmd/influx/write.go#L154

File: cmd/influx/main.go (hunk 2 above)
Comment (jsteenb2) [reply cmd/influx/main.go]: did the dryrun command get added? I'm not seeing it 😬

File: cmd/influx/write.go [hunk 7]
```diff
@@ -68,94 +62,133 @@ or add an entire file specified with an @ prefix.`
 		},
 	}
 	opts.mustRegister(cmd)
-
+	cmd.PersistentFlags().StringVar(&writeFlags.Format, "format", "", "Input format, either lp (Line Protocol) or csv (Comma Separated Values). Defaults to lp unless '.csv' extension")
+	cmd.PersistentFlags().StringVarP(&writeFlags.File, "file", "f", "", "The path to the file to import")
+
+	cmdDryRun := opt.newCmd("dryrun", func(command *cobra.Command, args []string) error {
```
Comment (jsteenb2) [inline cmd/influx/write.go]: can you break out the `dryrun` command into its own command?

File: cmd/influx/write.go [hunk 8]
```diff
@@ -68,94 +62,133 @@ or add an entire file specified with an @ prefix.`
...
-			InsecureSkipVerify: flags.skipVerify,
-		},
+	if writeFlags.Format == inputFormatCsv {
+		r = write.CsvToProtocolLines(r)
 	}
 
-	ctx = signals.WithStandardSignals(ctx)
-	if err := s.Write(ctx, orgID, bucketID, r); err != nil && err != context.Canceled {
-		return fmt.Errorf("failed to write data: %v", err)
+	if writeFlags.DryRun {
```
Comment (jsteenb2) [inline cmd/influx/write.go]: curiosity: feels like the Dry Run functionality can be pulled out from this large func. It shares some things, but those shared things you can pull out as needed and clean up the rest of this here. This command was already doing a fair amount, the dry run stuff baked into here makes it harder to follow. WDYT?

File: cmd/influx/write.go (hunk 7 above)
Comment (sranka) [reply cmd/influx/write.go]: Sure

File: cmd/influx/write.go (hunk 8 above)
Comment (sranka) [reply cmd/influx/write.go]: At the time when I was initially writing this, I thought that it is less error-prone to have a single function that is executed for both write and 'write with dry-run'. Now, after the changes, when the flags settle down and doing a few steps back to see the code better, I have to agree that the code complexity will reduce if there will be separate functions for each command that will call a shared function to obtain a line data reader. Good point, thx :)

**Comment by sanderson** [comment]
@sranka If using the linepart annotation in annotated CSV _without_ the datatype annotations, should the `time` columns should always be an integer (not RFC3339) according to the precision the user specifies (just like line protocol) or will it convert RFC3339 timestamps into an integer for you?

**Comment by sanderson** [comment]
@sranka Also, how do you handle string and boolean field values with the `linepart` annotation? Do you have to include a `datatype` annotation? Or is it smart enough to know?

**Comment by sranka** [comment]
@sanderson *linepart* annotation is now independent on *datatype* annotation, at least conceptually. linepart places column data in the line, *datatype* is just field data type. 

Regarding the data type annotation:
- if no data type is specified for a time column, it is detected from the data, so both number and RFC3339 string is supported, see https://github.com/influxdata/influxdb/blob/17003/writeFromCsv/write/csvTable.go#L349
- if no data type is specified for a field column, the column data are copied to line protocol as-is
- the datatype annotation for measurement and tag columns is ignored, the conversion to line protocol is given for these types

It might be simpler to introduce extra _field_,  _measurement_ or _tag_ (and _ignore_) values in the datatype annotation so that the user could then supply a single annotation line (_#datatype_) to specify fields (with types) and non-field columns (measurement, time, tag and ignored). As I write this comment I tend to change the existing implementation this way and remove the linepart annotation. WDYT?

There is a wide area of enhancements that could be done regarding CSV, once this gets merged: 
- time formats may vary
- header/annotation lines could be added from cli line without changing the CSV file
- constant tags, field values or measurement name could be supplied as well
- CSV with more time columns (such as https://data.humdata.org/dataset/novel-coronavirus-2019-ncov-cases) can be supported OOTB
- field types could be detected and enforced
- better error reporting, a better indication of progress / success / failure
- detection of data format without supplying a format flag
- ...
So this is just a first step to support CSV.

File: cmd/influx/write.go [hunk 9]
```diff
@@ -152,11 +184,27 @@ func fluxWriteF(cmd *cobra.Command, args []string) error {
...
-
 	ctx = signals.WithStandardSignals(ctx)
 	if err := s.Write(ctx, orgID, bucketID, r); err != nil && err != context.Canceled {
 		return fmt.Errorf("failed to write data: %v", err)
 	}
 
 	return nil
 }
+
+func fluxWriteDryrunF(cmd *cobra.Command, args []string) error {
```
Comment (jsteenb2) [inline cmd/influx/write.go]: excellent, CLI is looking good from my end. Thanks for running with feedback here 👍
